<?xml version="1.0"?>

<launch>
	<master auto="start"/>

	<arg name="uav_name" default="uav_0"/>
	<arg name="uav_id" default="0"/>

    <arg name="delta_t" default="0.01"/>
    <arg name="epsilon_1" default="0.10"/>
    <arg name="epsilon_2" default="0.0007"/>
    <arg name="kappa" default="0.1"/>

    <arg name="topic_pub_vel_cmd_common" default="/common/velocity"/>
    <arg name="topic_sub_states" default="scan"/>

	<rosparam param="/uwb_reloc_talk_node/mobsId"> [200, 201, 202]</rosparam>
	<rosparam param="/uwb_reloc_algo_node/albega"> [1.0, -0.29, 200]</rosparam>

 	<!-- UWB Communication -->
	<node respawn="true" pkg="uwb_reloc" type="uwb_talk_node" name="uwb_reloc_talk_node">
        <!-- configuration for the node -->
		<param name="nodeRate" type="int" value="50"/>
		<param name="restEnable" type="bool" value="true"/>
		<param name="ignrTimeoutUwbInit" type="bool" value="true"/>
		<param name="allZZero" type="bool" value="true"/>
		<param name="useViconVelocity" type="bool" value="false"/>

        <!-- configuration for the p4xx -->
		<param name="p4xxInterfaceChoice" type="int" value="1"/>
		<param name="p4xxSerialPort" type="string" value="200" />
		<param name="mode" type="int" value="0" />
		<param name="autoConfig" type="bool" value="false" />
		<param name="forceId" type="bool" value="true" />
		<param name="smFileName" value="/home/ubuntu/slotmaps/smrl3/200.csv" />
		<param name="rqstDataSize" type="int" value="16" />
		<param name="rspdDataSize" type="int" value="0" />

        <!-- configuration for fcc -->
		<param name="fccSerialPort" type="string" value="/dev/uwb" />
		<param name="fccBaudrate" type="int" value="921600" />

        <!-- configuration for comm. health -->
		<param name="logEnable" type="bool" value="true" />
		<param name="logInput" type="bool" value="true" />
		<param name="logOutput" type="bool" value="true" />
		<param name="syncFCC" type="bool" value="true" />

		<param name="topic_px4VelSub" type="string" value="/$(arg uav_name)/estimate/velocity" />
		<param name="topic_px4PosSub" type="string" value="/$(arg uav_name)/estimate/position" />
	</node>

  	<!-- Relative Localization - Estimation -->
	<node respawn="true" pkg="uwb_reloc" type="reloc_algo_node" name="uwb_reloc_algo_node" output="log">
		<param name="nodeRate" type="int" value="50"/>
		<param name="velNoise" type="double" value="0.2"/>
		<param name="initCovariance" type="double" value="0.0001"/>
		<param name="Rd" type="double" value="0.05"/>
		<param name="T" type="double" value="0.04"/>
		<param name="T" type="double" value="0.04"/>
		<param name="epsilon1" type="double" value="1.0"/>
		<param name="epsilon2" type="double" value="0.007"/>
		<param name="logEnable" type="bool" value="true" />
	</node>

  	<!-- Relative Localization - Control -->
    <node pkg="uwb_reloc_control" type="uwb_reloc_control_node" name="uwb_reloc_control_$(arg uav_name)"
        required="true" clear_params="true" output="screen">
        <param name="uav_name" value="$(arg uav_name)" />
        <param name="uav_id" value="$(arg uav_id)" />
        <param name="topic_sub_states" value="/rlForControl" />
        <param name="topic_pub_vel_ctrl" value="/$(arg uav_name)/cmd_vel" />
        <param name="topic_pub_vel_cmd_common" value="/common/cmd_vel" />
        <param name="delta_t" value="$(arg delta_t)" />
        <param name="epsilon_1" value="$(arg epsilon_1)" />
        <param name="epsilon_2" value="$(arg epsilon_2)" />
        <param name="kappa" value="$(arg kappa)" />
    </node>

  	<!-- Message Translator -->
    <node pkg="uwb_reloc_control" type="msg_translator_node" name="msg_translator_node"
        required="true" clear_params="true" output="screen">
        <param name="topic_sub_vel_command" value="/$(arg uav_name)/cmd_vel" />
        <param name="topic_sub_pos_update" value="/$(arg uav_name)/mavros/local_position/pose" />
        <param name="topic_sub_vel_update" value="/$(arg uav_name)/mavros/local_position/velocity" />

        <param name="topic_pub_pos_update" value="/$(arg uav_name)/estimate/position" />
        <param name="topic_pub_vel_update" value="/$(arg uav_name)/estimate/velocity" />
        <param name="topic_pub_vel_command" value="/$(arg uav_name)/mavros/setpoint_velocity/cmd_vel" />

        <param name="timeout_pose_update" value="1.0" />
        <param name="timeout_vel_update" value="1.0" />
        <param name="timeout_vel_command" value="1.0" />
    </node>

    <!-- MAVROS -->
    <node ns="$(arg uav_name)" pkg="mavros" type="mavros_node" name="mavros" 
    	required="true" clear_params="true" output="screen">
        <param name="fcu_url" value="/dev/px4:921600" />
        <param name="gcs_url" value="udp://14550@127.0.0.1:14556" />
        <param name="target_system_id" value="1" />
        <param name="target_component_id" value="1" />

        <rosparam command="load" file="$(find mavros)/launch/px4_pluginlists.yaml" />
        <rosparam command="load" file="$(find mavros)/launch/px4_config.yaml" />
    </node>

  	<!-- Recording -->
<!--   
	<node pkg="rosbag" type="record" name="uwb_talk_record" output="screen" 
		args="-o /home/ubuntu/drones/U0/U0_talk /uwb_talk"> 
	</node>
  	<node pkg="rosbag" type="record" name="reloc_est_record" output="screen" 
  		args="-o /home/ubuntu/drones/U0/U0_RlEst /uwb_RlEst"> 
  	</node> 
-->

</launch>
